### ISSUE TITLE
Calico Node Pod Failing - cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized

### ISSUE THREAD

**Issue by shashank-omre-cldcvr** [projectcalico/calico#8368]
Fresh Calico Operator(v3.27.0) Installation in 1.24 EKS cluster.

```
kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.27.0/manifests/tigera-operator.yaml
```

```
NAME                                       READY   STATUS                  RESTARTS        AGE
calico-kube-controllers-6bf98f54bb-bqtmq   1/1     Running                 0               15h
calico-node-47bvv                          0/1     Init:CrashLoopBackOff   167 (75s ago)   13h
calico-node-7wll8                          0/1     Init:CrashLoopBackOff   2 (23s ago)     70s
calico-node-br5qp                          0/1     Init:CrashLoopBackOff   6 (2m26s ago)   8m16s
calico-node-gln78                          0/1     Init:CrashLoopBackOff   8 (5m13s ago)   22m
calico-typha-5d4689887f-s9qkr              1/1     Running                 0               15h
calico-typha-5d4689887f-wdgdg              1/1     Running                 0               15h
csi-node-driver-26z8d                      2/2     Running                 0               15h
csi-node-driver-pzlhk                      2/2     Running                 0               15h
csi-node-driver-r4jtl                      2/2     Running                 0               71s
csi-node-driver-x8qn4                      2/2     Running                 0               15h
```

## Expected Behavior

All Pods should we be running state

## Current Behavior

calico-node pod throwing this Error.

```
2023-12-21 08:32:31.622 [ERROR][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
2023-12-21 08:32:31.622 [FATAL][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
```

## Steps to Reproduce (for bugs)
It's Fresh installation 
```
2023-12-21 08:32:31.622 [ERROR][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
2023-12-21 08:32:31.622 [FATAL][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
```

## Your Environment

* Calico version - v3.27.0
* Orchestrator version (e.g. kubernetes, mesos, rkt): EKS 1.24

**Comment by caseydavenport** [comment]
Looks like a potential RBAC issue - the cni-installer needs permissions to provision new tokens for the CNI plugin to use.

What does this show:

```
kubectl get clusterrole calico-node -o yaml
```

**Comment by jon-nfc** [comment]
Is there a fix for this? same problem, clean install using the operator. `kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.27.0/manifests/tigera-operator.yaml` I even tried the the master branch manifest, still to no avail. only version that works is non-operator install with manifest 3.25.0.

``` bash
root@7dd5f0bf32b32b93:/home/deploy# kubectl logs -f -n calico-system calico-node-wmdhj -c install-cni
2024-01-27 09:15:55.401 [INFO][1] cni-installer/<nil> <nil>: Running as a Kubernetes pod
2024-01-27 09:15:56.826 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/bandwidth"
2024-01-27 09:15:56.826 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/bandwidth
2024-01-27 09:15:56.955 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/calico"
2024-01-27 09:15:56.955 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/calico
2024-01-27 09:15:57.058 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/calico-ipam"
2024-01-27 09:15:57.058 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/calico-ipam
2024-01-27 09:15:57.063 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/flannel"
2024-01-27 09:15:57.063 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/flannel
2024-01-27 09:15:57.072 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/host-local"
2024-01-27 09:15:57.072 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/host-local
2024-01-27 09:15:57.202 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/install"
2024-01-27 09:15:57.202 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/install
2024-01-27 09:15:57.208 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/loopback"
2024-01-27 09:15:57.208 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/loopback
2024-01-27 09:15:57.215 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/portmap"
2024-01-27 09:15:57.215 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/portmap
2024-01-27 09:15:57.221 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/tuning"
2024-01-27 09:15:57.221 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/tuning
2024-01-27 09:15:57.221 [INFO][1] cni-installer/<nil> <nil>: Wrote Calico CNI binaries to /host/opt/cni/bin

2024-01-27 09:15:57.260 [INFO][1] cni-installer/<nil> <nil>: CNI plugin version: v3.27.0

2024-01-27 09:15:57.260 [INFO][1] cni-installer/<nil> <nil>: /host/secondary-bin-dir is not writeable, skipping
2024-01-27 09:15:57.260 [WARNING][1] cni-installer/<nil> <nil>: Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2024-01-27 09:15:57.284 [ERROR][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
2024-01-27 09:15:57.284 [FATAL][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
root@7dd5f0bf32b32b93:/home/deploy#
```

**Comment by caseydavenport** [comment]
For anyone encountering this issue, please run the following command to determine if your cluster's RBAC is correct:

```
kubectl auth can-i create serviceaccounts/calico-cni-plugin -n calico-system --subresource token --as "system:serviceaccount:calico-system:calico-node"
```

This should return `yes`, indicating that calico-node is allowed to create tokens for the CNI plugin. If this returns `yes`, then Calico's RBAC is configured correctly and the `Unauthorized` likely means something else - for example, bad certificates being provided to the containers.

If this returns `no`, that indicates that something is wrong with the RBAC in the cluster.

**Comment by Brice187** [comment]
```
15:38:15 lars@d04  ~ » devkubectl -n calico-system logs calico-node-7wvn8 -c install-cni
2024-02-16 14:38:11.600 [INFO][1] cni-installer/<nil> <nil>: Running as a Kubernetes pod
2024-02-16 14:38:11.620 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/bandwidth"
2024-02-16 14:38:11.620 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/bandwidth
2024-02-16 14:38:11.807 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/calico"
2024-02-16 14:38:11.807 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/calico
2024-02-16 14:38:11.977 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/calico-ipam"
2024-02-16 14:38:11.977 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/calico-ipam
2024-02-16 14:38:11.989 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/flannel"
2024-02-16 14:38:11.989 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/flannel
2024-02-16 14:38:12.005 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/host-local"
2024-02-16 14:38:12.005 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/host-local
2024-02-16 14:38:12.016 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/loopback"
2024-02-16 14:38:12.016 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/loopback
2024-02-16 14:38:12.030 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/portmap"
2024-02-16 14:38:12.031 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/portmap
2024-02-16 14:38:12.042 [INFO][1] cni-installer/<nil> <nil>: File is already up to date, skipping file="/host/opt/cni/bin/tuning"
2024-02-16 14:38:12.043 [INFO][1] cni-installer/<nil> <nil>: Installed /host/opt/cni/bin/tuning
2024-02-16 14:38:12.043 [INFO][1] cni-installer/<nil> <nil>: Wrote Calico CNI binaries to /host/opt/cni/bin

2024-02-16 14:38:12.098 [INFO][1] cni-installer/<nil> <nil>: CNI plugin version: v3.28.0-0.dev-386-g3a8d575515b1

2024-02-16 14:38:12.098 [INFO][1] cni-installer/<nil> <nil>: /host/secondary-bin-dir is not writeable, skipping
2024-02-16 14:38:12.098 [WARNING][1] cni-installer/<nil> <nil>: Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2024-02-16 14:38:12.112 [ERROR][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
2024-02-16 14:38:12.112 [FATAL][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=Unauthorized
15:38:23 lars@d04  ~ » devkubectl auth can-i create serviceaccounts/calico-cni-plugin -n calico-system --subresource token --as "system:serviceaccount:calico-system:calico-node"
yes

**Comment by Enquier** [comment]
I am getting a similar result as @Brice187 , using an Operator install on 3.27.0

**Comment by Enquier** [comment]
This issue is driving me nuts. It seems to happen consistently across all of my nodes

**Comment by caseydavenport** [comment]
The fact that authorization believes that the CNI plugin is authorized to create tokens suggests this is likely a problem with the certificates and/or tokens being used by the CNI plugin in order to make the request, rather than RBAC configuration itself.

I'd first check that your node clocks are properly set and synchronized - this can cause issues where a valid token or certificate can be treated as invalid if the authorizing node's clock is not set properly. 

Another case that I have seen this happen is if nodes or other state have been restored from a backup in some capacity, resulting in cached secrets, etc. that are no longer valid.

**Comment by caseydavenport** [comment]
Other things that would be interesting to see:

- Parse the JWT in the `/etc/cni/net.d/10-calico.conflist`  file to see if it looks valid, and what its expiry time is.
- See if you can find the JWT being used by calico-node and do the same. This might be trickier as I am not sure it's ever written to disk, but it might be possible.
- Enable verbose logging in the API server and see if it provides any more context as to why it's rejecting the requests.

**Comment by caseydavenport** [comment]
Is everyone encountering this issue running on EKS?

**Comment by Enquier** [comment]
There is no JWT in my file:

```
{
                          "name": "k8s-pod-network",
                          "cniVersion": "0.3.1",
                          "plugins": [{"container_settings":{"allow_ip_forwarding":false},"datastore_type":"kubernetes","ipam":{"assign_ipv4":"true","assign_ipv6":"false","type":"calico-ipam"},"kubernetes":{"k8s_api_root":"https://20.96.0.1:443","kubeconfig":"/etc/cni/net.d/calico-kubeconfig"},"log_file_max_age":30,"log_file_max_count":10,"log_file_max_size":100,"log_file_path":"/var/log/calico/cni/cni.log","log_level":"Info","mtu":0,"nodename_file_optional":false,"policy":{"type":"k8s"},"type":"calico"},{"capabilities":{"bandwidth":true},"type":"bandwidth"},{"capabilities":{"portMappings":true},"snat":true,"type":"portmap"}]
                        }
```

**Comment by caseydavenport** [comment]
> There is no JWT in my file:

D'oh, yes it should be in the kubeconfig.

That token has:

```
Issued at: Mon Mar 04 2024 23:24:23 GMT+0000
Expires:   Tue Mar 05 2024 23:24:23 GMT+0000
```

Which checks out, but unfortunately doesn't tell us much about the calico-node token other than that the calico-node token was valid at the "Issued at" time. The calico-node token is the one that is ultimately the one in question here. Not sure yet how to get access to that one.

The other avenue to explore here is if Kubenretes is having trouble refreshing the token it assigns to calico/node.

> Kubernetes gets a short-lived, automatically rotating token using the TokenRequest API and mounts the token as a [projected volume](https://kubernetes.io/docs/concepts/storage/projected-volumes/#serviceaccounttoken).

If there is an issue with the TokenRequest or with the projected volume it might manifest in this way. Kubernetes / kubelet / apiserver logs might show more.

**Comment by tomastigera** [comment]
Any update on this?

**Comment by clbx** [comment]
I also ran into this after a power outage, RBAC is valid. 
Running on Debian 12, Kubernetes 1.28.2, Calico 3.26.4 using Tigera Operator 1.30.9

**Comment by caseydavenport** [comment]
> Not sure why this is happening, but I think it has something to do with the operator keeping the service account alive while k8s api thinks its removed. As soon as I removed the finalizer from the svc acct, it was deleted and re-created

This is an awesome piece of the puzzle, thank you! 

Do you have any idea why the service account may have been terminating in the first place? I wouldn't expect a power outage to cause that.

There is a known issue preventing graceful termination of that service account in the case that the Installation resource is deleted, that will be fixed in v3.28. I wonder if that is related, or if the use of a finalizer on that serviceaccount more generally is related to this issue in other clusters. 

For anyone still encountering this issue, could you please show the output of this command as well? @Enquier @Brice187 @jon-nfc 

```
kubectl get serviceaccount -n calico-system calico-node -o yaml
```

**Comment by clbx** [comment]
> Do you have any idea why the service account may have been terminating in the first place?

No idea, unfortunately I don't have any logs from before the outage, I didn't get to it for a few hours and the api server logs don't stay for very long.

**Comment by DJanyavula** [comment]
I am also facing the same issue
```
2024-04-17 08:57:17.484 [INFO][1] cni-installer/<nil> <nil>: Wrote Calico CNI binaries to /host/opt/cni/bin

2024-04-17 08:57:17.596 [INFO][1] cni-installer/<nil> <nil>: CNI plugin version: v3.27.0

2024-04-17 08:57:17.596 [INFO][1] cni-installer/<nil> <nil>: /host/secondary-bin-dir is not writeable, skipping
2024-04-17 08:57:17.597 [WARNING][1] cni-installer/<nil> <nil>: Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2024-04-17 08:57:17.712 [ERROR][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=serviceaccounts "calico-cni-plugin" is forbidden: User "system:serviceaccount:kube-system:calico-node" cannot create resource "serviceaccounts/token" in API group "" in the namespace "kube-system"
2024-04-17 08:57:17.712 [FATAL][1] cni-installer/<nil> <nil>: Unable to create token for CNI kubeconfig error=serviceaccounts "calico-cni-plugin" is forbidden: User "system:serviceaccount:kube-system:calico-node" cannot create resource "serviceaccounts/token" in API group "" in the namespace "kube-system"
```

**Comment by caseydavenport** [comment]
@DJanyavula please read the earlier comments - there are a number of diagnostics requested for anyone hitting this issue, and without those we can't help.

**Comment by txbxxx** [comment]
root@Tc-Server:/opt/calio# kubectl auth can-i create serviceaccounts/calico-cni-plugin -n calico-system --subresource token --as "system:serviceaccount:calico-system:calico-node"
no
Using kind to build k8s, using calio when cni, this return no how to solve?

**Comment by caseydavenport** [comment]
```
kubectl auth can-i create serviceaccounts/calico-cni-plugin -n calico-system --subresource token --as "system:serviceaccount:calico-system:calico-node"
```

This command checks if the CNI plugin has the necessary permissions. If it does not, it means the ClusterRole, ClusterRoleBinding, or ServiceAccount are not correctly configured or installed on the cluster. 

For anyone experiencing this, please first verify that the Calico CNI plugin RBAC resources exist on the cluster:

```
kubectl get clusterrole calico-cni-plugin
kubectl get clusterrolebinding calico-cni-plugin
kubectl get serviceaccount -n calico-system calico-cni-plugin
```

If any of those **do not exist**, it means the tigera/operator is struggling to create them for some reason. Please look at the tigera-operator logs, and the output from `kubectl describe tigerastatus calico` to see what the issue might be.

If those resources _do_ exist, the next thing to check is whether they are stuck in a `Terminating` state or not. You can do this by checking the `deletionTimestamp` field on each of them.

```
kubectl get clusterrole calico-cni-plugin | grep deletionTimestamp
kubectl get clusterrolebinding calico-cni-plugin | grep deletionTimestamp
kubectl get serviceaccount -n calico-system calico-cni-plugin | grep deletionTimestamp
```

If you see a `deletionTimestmp` field on any of them, it means that the resource is trying to be deleted but is stuck. In this case, you need to edit each resource to remove the `cni-protector` entry from the `finalizers` field to allow the resources to terminate and the tigera-operator to re-create them. Once that is done, you can restart calico/node and things should start to function.

This is a result of a bug that is triggered often when deleting / re-creating the `Installation` resource, or manually deleting the RBAC resources by hand. The former has been fixed in v3.28, and the latter should be simply be avoided. If you _do_ need to delete these by hand without deleting the `Installation`, know that you will need to manually remove the finalizer and that they will be recreated by the operator.

**Comment by ajaypraj** [comment]
Hi All, I am facing same issue. I am trying to setup the cluster using kubernetes 1.26 and kubeadm . But calio pods are not coming up due to token creation failure .
Here is log for same:
```
{"log":"2024-06-05 12:38:15.352 [ERROR]. [1] cni-installer/\u003cnil\u003e \u003cnil\u003e: Unable to create token for CNI kubeconfig error=Post \"https://172.16.0.1:443/api/v1/namespaces/kube-system/serviceaccounts/calico-cni-plugin/token\": dial tcp 172.16.0.1:443: i/o timeout\n","stream":"stderr","time":"2024-06-05T12:38:15.352838536Z"} 

{"log":"2024-06-05 12:38:15.352 [FATAL][1] cni-installer/\u003cnil\u003e \u003cnil\u003e: Unable to create token for CNI kubeconfig error=Post \"https://172.16.0.1:443/api/v1/namespaces/kube-system/serviceaccounts/calico-cni-plugin/token\": dial tcp 172.16.0.1:443: i/o timeout\n","stream":"stderr","time":"2024-06-05T12:38:15.352956389Z"}

```
<img width="598" alt="image" src="https://github.com/projectcalico/calico/assets/34944605/1d92f75f-bf18-43c2-bec0-71f767790e8b">



 **Expected behaviour**
 All pods should be up and running .
 
  **Current behaviour**
 After running kubeadm init all pods are running expect core-dns pods . Installed calico cni using [manifest](https://raw.githubusercontent.com/projectcalico/calico/v3.26.4/manifests/calico.yaml)  .
 
Checked the service account and found that secret token is missing.

 root@debian:~# kubectl get serviceaccount -n kube-system calico-node -o yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: "2024-06-04T11:30:25Z"
  name: calico-node
  namespace: kube-system
  resourceVersion: "139365"
  uid: 7c3bdfb9-1073-4006-ac8c-61d619ed0a91
 

 Checked Calico CNI plugin RBAC resources exist on the cluster:
```
root@debian:~# kubectl get clusterrole calico-cni-plugin
NAME                CREATED AT
calico-cni-plugin   2024-06-04T11:30:25Z
root@debian:~# kubectl get clusterrolebinding calico-cni-plugin
NAME                ROLE                            AGE
calico-cni-plugin   ClusterRole/calico-cni-plugin   25h
root@debian:~# kubectl get serviceaccount -n kube-system calico-cni-plugin
NAME                SECRETS   AGE
calico-cni-plugin   0         25h
```

====
checked the resources are not stuck in terminating state

```
root@debian:~# kubectl get serviceaccount -n kube-system calico-cni-plugin
NAME                SECRETS   AGE
calico-cni-plugin   0         25h
root@debian:~# kubectl get clusterrole calico-cni-plugin | grep deletionTimestamp
root@debian:~# kubectl get clusterrolebinding calico-cni-plugin | grep deletionTimestamp
root@debian:~# kubectl get serviceaccount -n kube-system calico-cni-plugin | grep deletionTimestamp
```
however, running command can-i create service account it is returning no.
```
kubectl auth can-i create serviceaccounts/calico-cni-plugin -n calico-system --subresource token --as "system:serviceaccount:calico-system:calico-node"
no
```
 **How to reproduce it :**
 kubeadm init cluster_config.yaml 
kubectl apply -f calico.yaml [here is [link](https://raw.githubusercontent.com/projectcalico/calico/v3.26.4/manifests/calico.yaml) of file ]

Here is cluster_config.yaml file:
```
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
nodeRegistration:
  criSocket: "/var/run/cri-dockerd.sock"

---
apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
kubernetesVersion: "v1.26.15"
networking:
  podSubnet: "172.16.0.0/24,fde1::/64"
  serviceSubnet: "172.16.1.0/16,fde1::/112"
controllerManager:
  extraArgs:
    feature-gates: "IPv6DualStack=true"
apiServer:
  extraArgs:
    advertise-address: "172.16.2.1"
    tls-cipher-suites: "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"

---
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
tlsCipherSuites:
  - "TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
  - "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
  - "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
```
**Environment details** 
I am running cluster on debian 11 VM, which is running on vlware vcenter.Here are details:
**OS**: Debian 11
**Kubernetes Version**: 1.26.15
**kubeadm Version**: 1.26.15
**Docker Version**: 26.1.3
**Container Runtime**: cri-dockerd 0.3.14

**Steps tried:**
rm -rf /etc/cni/net.d/* && systemctl restart kubelet  , but it does not resolve the issue.

Any troubleshooting or help is highly appreciated.

**Comment by jon-nfc** [comment]
@ajaypraj 

> {"log":"2024-06-05 12:38:15.352 [ERROR]. [1] cni-installer/\u003cnil\u003e \u003cnil\u003e: Unable to create token for CNI kubeconfig error=Post \"https://172.16.0.1:443/api/v1/namespaces/kube-system/serviceaccounts/calico-cni-plugin/token\": dial tcp 172.16.0.1:443: i/o timeout\n","stream":"stderr","time":"2024-06-05T12:38:15.352838536Z"} 

what address is `172.16.0.1:443` attached to? as error `dial tcp 172.16.0.1:443: i/o timeout\n"` may be the issue and seperate to calico

other command that may assist in fault finding:

- `kubectl get endpoints -A`
- `iptables -nvL` pay particular attention to the chains prefixed with `KUBE-` i.e. `KUBE-FORWARD`
- `curl -k https://<service ip for kubernetes.default >:443/api` should return JSON stating unauthorized access. this is good as it shows api connectivity OK.

**Comment by caseydavenport** [comment]
> what address is 172.16.0.1:443 attached to? as error dial tcp 172.16.0.1:443: i/o timeout\n" may be the issue and seperate to calico

Yeah, I agree - @ajaypraj it sounds to me like you are encountering a different issue from this one. Could you raise a separate GitHub issue to track that?

**Comment by ajaypraj** [comment]
@caseydavenport sure.

**Comment by acrjene** [comment]
I fixed this problem by disabling VPN

**Comment by oniubi** [comment]
[root@k8s-master01 cfg]# kubectl logs calico-node-c74kn -n kube-system -c install-cni
2024-09-10 09:48:43.661 [INFO][1] cni-installer/install.go 142: Running as a Kubernetes pod
2024-09-10 09:48:43.667 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/bandwidth"
2024-09-10 09:48:43.673 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/bandwidth
2024-09-10 09:48:43.739 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/calico"
2024-09-10 09:48:43.787 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/calico
2024-09-10 09:48:43.843 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/calico-ipam"
2024-09-10 09:48:43.893 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/calico-ipam
2024-09-10 09:48:43.896 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/flannel"
2024-09-10 09:48:43.898 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/flannel
2024-09-10 09:48:43.901 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/host-local"
2024-09-10 09:48:43.904 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/host-local
2024-09-10 09:48:43.908 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/loopback"
2024-09-10 09:48:43.912 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/loopback
2024-09-10 09:48:43.917 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/portmap"
2024-09-10 09:48:43.920 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/portmap
2024-09-10 09:48:43.924 [INFO][1] cni-installer/install.go 419: File is already up to date, skipping file="/host/opt/cni/bin/tuning"
2024-09-10 09:48:43.927 [INFO][1] cni-installer/install.go 228: Installed /host/opt/cni/bin/tuning
2024-09-10 09:48:43.927 [INFO][1] cni-installer/install.go 232: Wrote Calico CNI binaries to /host/opt/cni/bin

2024-09-10 09:48:43.927 [INFO][1] cni-installer/install.go 237: CNI plugin version: v3.28.1
2024-09-10 09:48:43.927 [INFO][1] cni-installer/install.go 189: /host/secondary-bin-dir is not writeable, skipping
2024-09-10 09:48:43.927 [WARNING][1] cni-installer/winutils.go 150: Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2024-09-10 09:48:43.934 [ERROR][1] cni-installer/token_watch.go 108: Unable to create token for CNI kubeconfig error=Unauthorized
2024-09-10 09:48:43.934 [FATAL][1] cni-installer/install.go 482: Unable to create token for CNI kubeconfig error=Unauthorized

**Comment by caseydavenport** [comment]
@oniubi please see this comment: https://github.com/projectcalico/calico/issues/8368#issuecomment-2120873448

If that doesn't help, then please raise a new issue.

**Comment by oniubi** [comment]
[root@k8s-master01 cfg]# kubectl get ns
NAME              STATUS   AGE
default           Active   20h
kube-node-lease   Active   20h
kube-public       Active   20h
kube-system       Active   20h
[root@k8s-master01 cfg]#
[root@k8s-master01 cfg]# kubectl create -f tigera-operator.yaml
namespace/tigera-operator created
customresourcedefinition.apiextensions.k8s.io/bgpconfigurations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/bgpfilters.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/bgppeers.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/blockaffinities.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/caliconodestatuses.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/clusterinformations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/felixconfigurations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/globalnetworkpolicies.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/globalnetworksets.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/hostendpoints.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamblocks.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamconfigs.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipamhandles.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ippools.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/ipreservations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/kubecontrollersconfigurations.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/networkpolicies.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/networksets.crd.projectcalico.org created
customresourcedefinition.apiextensions.k8s.io/apiservers.operator.tigera.io created
customresourcedefinition.apiextensions.k8s.io/imagesets.operator.tigera.io created
customresourcedefinition.apiextensions.k8s.io/installations.operator.tigera.io created
customresourcedefinition.apiextensions.k8s.io/tigerastatuses.operator.tigera.io created
serviceaccount/tigera-operator created
clusterrole.rbac.authorization.k8s.io/tigera-operator created
clusterrolebinding.rbac.authorization.k8s.io/tigera-operator created
deployment.apps/tigera-operator created
[root@k8s-master01 cfg]# kubectl get ns
NAME              STATUS   AGE
default           Active   20h
kube-node-lease   Active   20h
kube-public       Active   20h
kube-system       Active   20h
tigera-operator   Active   8s
[root@k8s-master01 cfg]# kubectl get pods -n tigera-operator
NAME                               READY   STATUS    RESTARTS   AGE
tigera-operator-5d56685c77-tzhv2   1/1     Running   0          22s
[root@k8s-master01 cfg]#
[root@k8s-master01 cfg]# kubectl logs tigera-operator-5d56685c77-tzhv2 -n tigera-operator
2024/09/14 03:49:25 [INFO] Version: v1.34.3
2024/09/14 03:49:25 [INFO] Go Version: go1.22.5 X:boringcrypto
2024/09/14 03:49:25 [INFO] Go OS/Arch: linux/amd64
W0914 03:49:25.660142       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:25.660162       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
W0914 03:49:26.954102       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:26.954134       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
W0914 03:49:29.741328       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:29.741373       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
W0914 03:49:34.141151       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:34.141202       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
W0914 03:49:40.663507       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:40.663553       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized
W0914 03:49:58.678254       1 reflector.go:535] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: failed to list *v1.ConfigMap: Unauthorized
E0914 03:49:58.678277       1 reflector.go:147] pkg/mod/k8s.io/client-go@v0.28.9/tools/cache/reflector.go:229: Failed to watch *v1.ConfigMap: failed to list *v1.ConfigMap: Unauthorized

[root@k8s-master01 cfg]# kubectl auth can-i list configmaps -n tigera-operator
yes
[root@k8s-master01 cfg]# kubectl auth can-i watch configmaps -n tigera-operator
yes
[root@k8s-master01 cfg]# kubectl get clusterrole tigera-operator -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: "2024-09-14T03:49:24Z"
  name: tigera-operator
  resourceVersion: "58085"
  uid: 562e414f-7d14-46fd-9867-8ac649930e67
rules:
- apiGroups:
  - ""
  resources:
  - namespaces
  - pods
  - podtemplates
  - services
  - endpoints
  - events
  - configmaps
  - secrets
  - serviceaccounts
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
- apiGroups:
  - ""
  resources:
  - resourcequotas
  verbs:
  - list
  - get
  - watch
- apiGroups:
  - ""
  resourceNames:
  - calico-critical-pods
  - tigera-critical-pods
  resources:
  - resourcequotas
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
- apiGroups:
  - ""
  resources:
  - nodes
  verbs:
  - get
  - patch
  - list
  - watch
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - clusterroles
  - clusterrolebindings
  - rolebindings
  - roles
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
  - bind
  - escalate
- apiGroups:
  - apps
  resources:
  - deployments
  - daemonsets
  - statefulsets
  verbs:
  - create
  - get
  - list
  - patch
  - update
  - delete
  - watch
- apiGroups:
  - apps
  resourceNames:
  - tigera-operator
  resources:
  - deployments/finalizers
  verbs:
  - update
  - list
  - watch
- apiGroups:
  - operator.tigera.io
  resources:
  - apiservers/finalizers
  - installations
  - installations/status
  - installations/finalizers
  - tigerastatuses
  - tigerastatuses/status
  - tigerastatuses/finalizers
  - apiservers
  - apiservers/status
  - imagesets
  verbs:
  - get
  - list
  - update
  - patch
  - watch
- apiGroups:
  - operator.tigera.io
  resources:
  - tigerastatuses
  verbs:
  - create
  - delete
  - list
  - watch
- apiGroups:
  - operator.tigera.io
  resources:
  - installations
  - apiservers
  verbs:
  - delete
  - list
  - watch
- apiGroups:
  - networking.k8s.io
  resources:
  - networkpolicies
  verbs:
  - create
  - update
  - delete
  - get
  - list
  - watch
- apiGroups:
  - crd.projectcalico.org
  resources:
  - felixconfigurations
  - ippools
  verbs:
  - create
  - patch
  - list
  - get
  - watch
- apiGroups:
  - crd.projectcalico.org
  resources:
  - kubecontrollersconfigurations
  - bgpconfigurations
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - projectcalico.org
  resources:
  - ippools
  verbs:
  - create
  - update
  - delete
  - patch
  - get
  - list
  - watch
- apiGroups:
  - projectcalico.org
  resources:
  - ipamconfigurations
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - scheduling.k8s.io
  resources:
  - priorityclasses
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
- apiGroups:
  - policy
  resources:
  - poddisruptionbudgets
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
- apiGroups:
  - apiregistration.k8s.io
  resources:
  - apiservices
  verbs:
  - list
  - watch
  - create
  - update
- apiGroups:
  - coordination.k8s.io
  resources:
  - leases
  verbs:
  - create
  - get
  - list
  - update
  - delete
  - watch
- apiGroups:
  - storage.k8s.io
  resources:
  - csidrivers
  verbs:
  - list
  - watch
  - update
  - get
  - create
  - delete
- apiGroups:
  - certificates.k8s.io
  resources:
  - certificatesigningrequests
  verbs:
  - list
  - watch
- apiGroups:
  - policy
  resourceNames:
  - tigera-operator
  resources:
  - podsecuritypolicies
  verbs:
  - use
- apiGroups:
  - policy
  resources:
  - podsecuritypolicies
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - delete

**Comment by sanadhis** [comment]
Facing the same issue after accidentally deleting `calico-system` namespace. The namespace was stuck in `Terminating` phase and I forcefully deleted it.

After, tigera-operator tried to re-create everything but the old service account `calico-node` remains (even after namespace was wiped out at some point).
Then calico-node pods were created but facing the same error: `Unable to create token for CNI kubeconfig error=Unauthorized`

What I did just to remove the previous service account:
```
kubectl -n calico-system patch sa calico-node -p '{"metadata": {"finalizers": null}}'
kubectl -n calico-system delete sa calico-node
```
Then it works like a charm as tigera-operator then able to re-create the service account.

exactly as @caseydavenport mentioned on https://github.com/projectcalico/calico/issues/8368#issuecomment-2120873448

> if you see a deletionTimestmp field on any of them, it means that the resource is trying to be deleted but is stuck. In this case, you need to edit each resource to remove the cni-protector entry from the finalizers field to allow the resources to terminate and the tigera-operator to re-create them

**Comment by Deepikasrini007** [comment]
Hi @caseydavenport 
i have checked it, i dont have deletionTimestamp in calico-cni-plugin. but still am getting this error in coredns pods. and the status of the pod is containercreating,

plugin type="calico" failed (add): error getting ClusterInformation: Get "https://172.20.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations/default": Service Unavailable

Please help!!
